app/services/ValidatorPool.scala

```diff
@@ -1,67 +1,121 @@
 package services
 
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.{ConcurrentHashMap}
 
 import scala.collection.JavaConverters._
-import scala.concurrent.{ExecutionContext, Future, Promise, blocking}
-
+import scala.concurrent.{ExecutionContext, Future, Promise}
 import play.api.Logger
-
-import model.Rule
+import model.{Category, Rule, RuleMatch}
 import utils.Validator
 import utils.Validator._
 
-case class ValidatorConfig(rules: List[Rule])
+import scala.util.{Failure, Success}
+import akka.stream.QueueOfferResult.{Dropped, Failure => QueueFailure, QueueClosed}
+import akka.stream._
+import akka.stream.scaladsl.{Sink, Source}
 
+case class ValidatorConfig(rules: List[Rule])
 case class ValidatorRequest(category: String, text: String)
 
-trait ValidatorFactory {
-  def createInstance(name: String, config: ValidatorConfig)(implicit ec: ExecutionContext): (Validator, List[String])
-}
+class ValidatorPool(val maxCurrentJobs: Int = 4, val maxQueuedJobs: Int = 100)(implicit ec: ExecutionContext, implicit val mat: Materializer) {
+  type TValidateCallback = () => Future[ValidatorResponse]
 
-class ValidatorPool(factory: ValidatorFactory)(implicit ec: ExecutionContext) {
+  case class ValidationJob(id: String, categoryId: String, text: String, promise: Promise[ValidatorResponse])
 
-  private val validators = new ConcurrentHashMap[String, Promise[Validator]]().asScala
+  private val validators = new ConcurrentHashMap[String, (Category, Validator)]().asScala
+  private val flow = Source.queue[ValidationJob](maxQueuedJobs, OverflowStrategy.dropNew)
```

Comment:
Definitely not against using Akka Streams here, the code is very clear. Just wondering what more it offers compared to use a Thread Pool and then native Java concurrent `BoundedQueue.offer` [methods](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ArrayBlockingQueue.html#offer-E-)?

Reply:
As discussed IRL -- we get backpressure and concurrency limits, but there's nothing to stop us implementing those things via a queue or similar. I'm inclined to use an abstraction that provides those things as part of its model rather than roll our own thing, but if there are good reasons to avoid Akka then I'm all 👂!